src/variant.h

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2017-2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_VARIANT_H
+#define BITCOIN_VARIANT_H
+
+#include <utility>
+
+#include <boost/variant.hpp>
+
+//! Substitute for C++17 std::variant
+template <class... Types>
+class Variant
+{
+private:
+    boost::variant<Types...> m_var;
+
+public:
+    Variant()
+    {
+        m_var = {};
+    }
+    Variant(const Variant& other)
+    {
+        m_var = other.m_var;
+    }
+    Variant(Variant& other)
+    {
+        m_var = other.m_var;
+    }
```

Comment:
In commit "Add a new Variant type" (85882c01d20add49f7b1f8b741e078968ec8080b)

Maybe drop this overload. I don't think it's doing anything the other two overloads don't cover. Also std::variant doesn't have this: https://en.cppreference.com/w/cpp/utility/variant/variant

Reply:
You're right, I was sure I dropped it.
There's no reason for both const and non-const constructor if it doesn't mutate.

Reply:
I see now why I didn't drop it :)
Because then the following code won't compile:
```
Variant<int> var = 5;
Variant<int> a = var;
```

Because `Variant& operator=(T&& value)` signature can match it "better" then `Variant(const Variant& other)` so it will choose that and then it will try to handle it like a value instead of a variant  which will fail to compile (https://gcc.godbolt.org/z/BKzBkE) not quite sure how libstd makes it work (I think it can be done with `std::enable_if` but that's uglier than another constructor)

Reply:
> I see now why I didn't drop it :)

Makes sense, I see why it's there now

Reply:
How would you feel about removing this and adding `enable_if` to all the templated versions that checks its not of type Variant?

Reply:
> How would you feel about removing this and adding `enable_if` to all the templated versions that checks its not of type Variant?
> 
> The tradeoff here is more specialized constructors versus less readable templates

Either way seems good to me. You should choose which way you like best / hate least. I guess it's a choice between defining more repetitive constructors, or being more concise but using enable_if.

Reply:
Done